contracts/distribution/PushDistributor.sol

```diff
@@ -0,0 +1,84 @@
+// SPDX-License-Identifier: GPL-3.0
+pragma solidity 0.8.20;
+
+import { IERC20 } from "../tokens/IERC20.sol";
+import { LibSafeCast } from "../utils/LibSafeCast.sol";
+import { LibERC20Compat } from "./../utils/LibERC20Compat.sol";
+import { LibAddress } from "./../utils/LibAddress.sol";
+
+// TODO: Add tests
+// TODO: Add events
+// TODO: Add custom errors
+
+interface IParty {
+    struct GovernanceValues {
+        uint40 voteDuration;
+        uint40 executionDelay;
+        uint16 passThresholdBps;
+        uint96 totalVotingPower;
+    }
+
+    function getGovernanceValues() external view returns (GovernanceValues memory);
+    function getIntrinsicVotingPowerAt(
+        address member,
+        uint256 timestamp,
+        uint256 hintIndex
+    ) external view returns (uint96);
+}
+
+contract PushDistributor {
+    using LibSafeCast for uint256;
+    using LibERC20Compat for IERC20;
+    using LibAddress for address payable;
+
+    // Token address used to indicate ETH.
+    address private constant ETH_ADDRESS = 0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE;
+
+    function distribute(IERC20 token, address[] memory members, uint256 amount) external payable {
+        IParty party = IParty(payable(msg.sender));
+        if (address(token) == ETH_ADDRESS) {
+            require(msg.value == amount, "Incorrect ETH amount");
+        } else {
+            require(msg.value == 0, "Unexpected ETH amount");
+            token.compatTransferFrom(msg.sender, address(this), amount);
+        }
+
+        address prevMember;
+        uint96 totalIntrinsicVotingPower;
+        uint96 totalVotingPower = party.getGovernanceValues().totalVotingPower;
```

Comment:
This doesn't work because it uses `address[] members` when the distribution was created but checks `totalVotingPower` at the time it was executed. What if we passed in `totalVotingPower` as a parameter at the time the proposal was created? Yes, a proposer could lie and give a wrong `totalVotingPower` and leave out some members from a distribution. But the same mechanisms we have against malicious proposals apply to prevent this (i.e. it can be vetoed, or simply won't pass).